src/awkward/_v2/operations/structure/ak_concatenate.py

```diff
@@ -4,524 +4,201 @@
-#                     [ak._v2.index.Index64(x) for x in all_counts],
-#                 )
-#                 inner = ak._v2.contents.UnionArray8_64(tags, index, all_flatten)
-
-#                 out = ak._v2.contents.ListOffsetArray64(
-#                     offsets, inner.simplify(merge=merge, mergebool=mergebool)
-#                 )
-#                 return lambda: (out,)
-
-#             elif any(
-#                 x.minmax_depth == (1, 1)
-#                 for x in inputs
-#                 if isinstance(x, ak._v2.contents.Content)
-#             ):
-#                 raise ValueError(
-#                     "at least one array is not deep enough to concatenate at "
-#                     "axis={0}".format(axis)
-#                 )
-
-#             else:
-#                 return None
-
-#         out = ak._v2._util.broadcast_and_apply(
-#             contents,
-#             getfunction,
-#             behavior=ak._v2._util.behaviorof(*arrays, behavior=behavior),
-#             allow_records=True,
-#             right_broadcast=False,
-#             pass_depth=True,
-#         )[0]
 
-#     return ak._v2._util.maybe_wrap(
-#         out, ak._v2._util.behaviorof(*arrays, behavior=behavior), highlevel
-#     )
+    #     """
+    #     Args:
+    #         arrays: Arrays to concatenate along any dimension.
+    #         axis (int): The dimension at which this operation is applied. The
+    #             outermost dimension is `0`, followed by `1`, etc., and negative
+    #             values count backward from the innermost: `-1` is the innermost
+    #             dimension, `-2` is the next level up, etc.
+    #         merge (bool): If True, combine data into the same buffers wherever
+    #             possible, eliminating unnecessary #ak.layout.UnionArray8_64 types
+    #             at the expense of materializing #ak.layout.VirtualArray nodes.
+    #         mergebool (bool): If True, boolean and nummeric data can be combined
+    #             into the same buffer, losing information about False vs `0` and
+    #             True vs `1`; otherwise, they are kept in separate buffers with
+    #             distinct types (using an #ak.layout.UnionArray8_64).
+    #         highlevel (bool): If True, return an #ak.Array; otherwise, return
+    #             a low-level #ak.layout.Content subclass.
+    #         behavior (None or dict): Custom #ak.behavior for the output array, if
+    #             high-level.
+
+    #     Returns an array with `arrays` concatenated. For `axis=0`, this means that
+    #     one whole array follows another. For `axis=1`, it means that the `arrays`
+    #     must have the same lengths and nested lists are each concatenated,
+    #     element for element, and similarly for deeper levels.
+    #     """
+
+    contents = [
+        ak._v2.operations.convert.to_layout(
+            x, allow_record=False if axis == 0 else True, allow_other=True
+        )
+        for x in arrays
+    ]
```

Comment:
This is where all inputs have been normalized. From this point on, we can assume that they're not raw NumPy arrays anymore (oy Python lists, etc.).